Update contracts release documentation (#17719)
ethereum-optimism/optimism: packages/contracts-bedrock/book/src/policies/release-process.md

# Tagging and Release Process

## Creating a tagged release

First select a tag string based on the guidance in [Monorepo Contracts Release Versioning](./versioning.md#monorepo-contracts-release-versioning)

1. Checkout the commit
1. Run `git tag <tag-string>`

> 1. Run `git push origin <tag-string>`

Reviewer: Although this renders as 1,2,3; was it intentional to have 1,1,1,... in the markdown source?

Reviewer 2: I believe it was intentional, it's a nice trick to minimize diffs when editing markdown

Author: Yeah, I find this makes it easier during the writing process, but probably better to set real numbers once the content is stable. 

```suggestion
1. Checkout the commit
2. Run `git tag <tag-string>`
3. Run `git push origin <tag-string>`
```

   - [standard-versions-sepolia.toml](https://github.com/ethereum-optimism/superchain-registry/blob/main/validation/standard/standard-versions-sepolia.toml)
1. Once the changes are merged into the superchain-registry, you can follow the [instructions](https://devdocs.optimism.io/op-deployer/reference-guide/releases.html#step-3-update-the-sr-with-the-new-release)
   for creating a new release of `op-deployer`.

## Implications for audits

The process above should be followed to create an `-rc.1` release prior to audit. This will be the target commit for
the audit. If any fixes are required by the audit results an Additional Release Candidate will be required.

## Additional Release Candidates

Sometimes fixes or additional changes need to be added to a release candidate version. In that case
we want to ensure fixes are made on both the release and the trunk branch, without stopping development
efforts on the trunk branch.

The process is as follows:

1. Make the fixes on `develop`. Increment the contracts semver as normal.
1. Create a new release branch, named `proposal/op-contracts/vX.Y.Z` off of the rc tag (all subsequent `-rc` tags
   will be made from this branch).

> 1. Cherry pick the fixes from `develop` into the release branch, and increment the semver as normal. If this increment results in any contract's semver being equal to or greater than it is on `develop`, then the semver should immediately be increased on `develop` to be greater than on the release branch.

Reviewer: This isn't super clear to me. If `develop` didn't progress _at all_ between one rc and the next, wouldn't we expect the patch, when applied to both `develop` and `proposal/op-contracts/vX.Y.Z` to result in identical semvers on the two branches? Why is that a problem? I don't really understand why the semver on `develop` needs to be strictly greater than it is on the `rc` (after all, this won't be the case when the first `rc` is cut).

Reviewer 2: This is basically a requirement that derives from the "always be release ready" principle. If we were to cut a new release from develop while the current release is still in progress (e.g. cut a contracts v6.0.0 on develop while v5.0.0 is going through audit or gov) for any contract in  v6 that was changed since the v5 rc, the v6 one should have a larger semver

Author: Based on a conversation held offline, I think we can resolve this by adding a bit of text to explain the rationale: 

```suggestion
1. Cherry pick the fixes from `develop` into the release branch, and increment the semver as normal. If this increment results in any contract's semver being equal to or greater than it is on `develop`, then the semver should immediately be increased on `develop` to be greater than on the release branch. This is necessary to avoid a situation where a given contract has two different implementations with the same version.
```

> 1. Cherry pick the fixes from `develop` into the release branch, and increment the semver as normal. If this increment results in any contract's semver being equal to or greater than it is on `develop`, then the semver should immediately be increased on `develop` to be greater than on the release branch. This is necessary to avoid a situation where a given contract has two different implementations with the same version.

Reviewer: ```suggestion
1. Cherry pick the fixes from `develop` into the release branch, and increment the semver as normal. If this increment results in any of the modified contracts' semver being equal to or greater than it is on `develop`, then the semver should immediately be increased on `develop` to be greater than on the release branch. This avoids a situation where a given contract has two different implementations with the same version.
```

I think what was tripping me up before was having _all_ contracts in scope for this semver check and not just the modified ones. We would expect unmodified contracts to have the same semver as develop immediately after the release branch is made, and I assume we do not want to modify the semver of those contracts on develop despite them being equal on the release branch.

Also suggesting to drop "necessary".

# Tagging and Release Process

## Creating a tagged release

First select a tag string based on the guidance in [Monorepo Contracts Release Versioning](./versioning.md#monorepo-contracts-release-versioning)

1. Checkout the commit
1. Run `git tag <tag-string>`
1. Run `git push origin <tag-string>`

>    This may require special authorization. Once pushed a tag cannot be deleted, so please be sure it is correct.

Reviewer 2: nit just to provide some more info
```suggestion
   Repo [rules](https://github.com/ethereum-optimism/optimism/rules/8196346?ref=refs%2Ftags%2Fop-contracts) require this is done by someone who is a [release-manager](https://github.com/orgs/ethereum-optimism/teams/release-managers). Once pushed a tag cannot be deleted, so please be sure it is correct.
```

   - [standard-versions-mainnet.toml](https://github.com/ethereum-optimism/superchain-registry/blob/main/validation/standard/standard-versions-mainnet.toml)
   - [standard-versions-sepolia.toml](https://github.com/ethereum-optimism/superchain-registry/blob/main/validation/standard/standard-versions-sepolia.toml)
1. Once the changes are merged into the superchain-registry, you can follow the [instructions](https://devdocs.optimism.io/op-deployer/reference-guide/releases.html#step-3-update-the-sr-with-the-new-release)
   for creating a new release of `op-deployer`.

## Implications for audits

The process above should be followed to create an `-rc.1` release prior to audit. This will be the target commit for
the audit. If any fixes are required by the audit results an Additional Release Candidate will be required.

## Additional Release Candidates

Sometimes fixes or additional changes need to be added to a release candidate version. In that case
we want to ensure fixes are made on both the release and the trunk branch, without stopping development
efforts on the trunk branch.

The process is as follows:

1. Make the fixes on `develop`. Increment the contracts semver as normal.

> 1. Create a new release branch, named `proposal/op-contracts/X.Y.Z` off of the rc tag (all subsequent `-rc` tags

Reviewer 2: ```suggestion
1. Create a new release branch, named `proposal/op-contracts/vX.Y.Z` off of the rc tag (all subsequent `-rc` tags
```

Author: oh, good catch!

   will be made from this branch).
1. Cherry pick the fixes from `develop` into that branch. Instead of incrementing the semver as normal,
   append `-patch.n` to the end of the version number. The value of `n` should start at 1 and be
   incremented for each additional patch. This suffix prevents a situation where two different implementations of the same

>    contract have the same semver.

Reviewer 2: For U16a, we instead incremented semver like normal on the release branch during the cherry-picks, and then once the release was done opened a [PR](https://github.com/ethereum-optimism/optimism/pull/17378) to develop to fix any semvers that were lower on develop than their counterpart on the release branch. What do you think of updating the guidance to use this process instead? This way we never deploy `.patch-` versions to productions to prod. 

If we wanted, we could also add a CI check ensure all contract semvers are greater than or equal to what they are on the latest release, that way you can't forget to update develop

Author: ```suggestion
1. Cherry pick the fixes from `develop` into the release branch, and increment the semver as normal. If this increment results in any contract's semver being equal to or greater than it is on `develop`, then the semver should immediately be increased on `develop` to be greater than on the release branch.
```

Author: I like it, incorporated that.

1. After merging the changes into the new release branch, tag the resulting commit on the proposal branch as `op-contracts/vX.Y.Z-rc.n`.
   Create a new release for this tag per the instructions above.

## Finalizing a release

Once a release has passed governance, a new tag should be created without the `-rc.n` suffix. To do this follow the
instructions in "Creating a tagged release" once again. It should not be necessary to redeploy the contracts with `op-deployer`,
but a new entry will be required in the superchain-registry's toml files regardless.

> When creating release notes, _uncheck_ the `Set as a pre-release`  option, and _check_ the
>    `Set as the latest release` option.

Reviewer 2: I'd recommend still unchecking this—I don't think we've every checked it in the past, and partners look for latest releases of client/go code more often than contract code, so it's more useful for them to see the client code as the "latest release" on the repo home page 
<img width="243" height="129" alt="image" src="https://github.com/user-attachments/assets/fedf4f8a-5a21-476f-8fa6-63a8fefb6da7" />

```suggestion
When creating release notes, _uncheck_ the `Set as a pre-release`  option, and _uncheck_ the
   `Set as the latest release` option.
```

Author: Good call. Adding some clarification on top of your diff:

```suggestion
When creating release notes, _uncheck_ the `Set as a pre-release`  option, and _uncheck_ the
   `Set as the latest release` option (latest releases are reserved for non-contract packages). 
```
